sdk-development-guidelines.md

```diff
@@ -0,0 +1,200 @@
+    * [Google Java Style](https://google.github.io/styleguide/javaguide.html)
+    * [Swift](https://github.com/raywenderlich/swift-style-guide)
+    * [Objective-C](https://github.com/raywenderlich/objective-c-style-guide)
+* Existing modules SHOULD follow a coding style guide. Always strive for **consistency** within a project's codebase
+* SHOULD use static code analysis, such as [detekt](https://github.com/arturbosch/detekt), clang-format, [SwiftLint](https://github.com/realm/SwiftLint), or oclint to minimise errors and minimise debates and formatting fix-up commits in PRs. It is RECOMMENDED to integrate these tools with the CI workflow as pre-commit git hooks.
+* SHOULD use quality tools available in common config: [Android](https://github.com/rakutentech/android-buildconfig) and [iOS](https://github.com/rakutentech/ios-buildconfig)
+* For Android <img src="images/android.png" alt="drawing" width="15"/>:
+    * MUST use the [Android Support Annotations](https://developer.android.com/studio/write/annotations) for public APIs
+    * Non-Android Java code MUST use the [Jetbrains Annotations](https://www.jetbrains.com/help/idea/annotating-source-code.html) for public APIs
+
+### Library Design
+
+#### Limit impact on consumers
+* SHOULD limit dependencies as much as possible primarily due to risk of version clashes e.g. SHOULD NOT depend on external networking library but also because the dependency may stop being maintained
+* SHOULD NOT have exported dependency on other SDK modules
+* SHOULD NOT have an exported dependency on other SDK modules e.g. if your module needs authentication, do not depend on Authentication SDK. Instead, require your library consumer to pass an access token string to your API instead.
+* SHOULD NOT have unnecessary exported dependencies, in particular
+    * SHOULD NOT leak networking library to consumers
+    * SHOULD NOT leak serialization library to consumers
+    * An exception to this rule are optional adapters for popular libraries
+* For Android <img src="images/android.png" alt="drawing" width="15"/>:
+    * MUST use a [consumerProguardFiles](https://google.github.io/android-gradle-dsl/current/com.android.build.gradle.internal.dsl.BuildType.html#com.android.build.gradle.internal.dsl.BuildType:consumerProguardFiles) file if obfuscation breaks the libraries function
+
+#### Expose minimal Surface area
+* Keep everything private unless it needs to be public. Every public class, method, property & resource MUST have a strong reason for being public
+* MAY keep internal source code in a `/Private` folder to allow tools to more easily ignore those files
+* MUST hide classes that need to be public for technical reasons but are not intended for SDK clients from documentation
```

Comment:
Wording is a bit strange. This may be better:
> MUST hide classes from generated API documentation that are required to be public solely for technical reasons and are not intended for SDK consumers

Reply:
changed to suggested statement